src/client/models/templates/Template.ts

```diff
@@ -56,6 +56,30 @@ export interface Templates {
     ]
 }
 
+export class TemplatesPushRequest {
+    SourceServerID: number;
+    DestinationServerID: number;
+    PerformChanges: boolean;
+
+    constructor(SourceServerID: number, DestinationServerID: number, PerformChanges: boolean) {
```

Comment:
@ibalosh Wy do we need a class for `TemplatesPushRequest`? I think an interface would be more idiomatic here. You wouldn't have to instantiate a class when calling `pushTemplates`, just pass options like:

```typescript
accountClient.pushTemplates({ SourceServerID: 1, DestinationServerID: 2, PerformChanges: false })
```

Reply:
that was our initial route when we moved to Typescript. However, @atheken suggested we moved to classes in the end for all calls that require sending a request. I think it was a good suggestion.

Having a class instead of interface gives you the benefit of being able to create an object, with faster error detection and clearer intelisense benefits of TS language.

Also, you can still create objects on the fly just like you would if you would use an interface. You don't need to instatiate an object if you don't want to. Call you provided would still work with classes too.

Reply:
I don't see any benefit of using a class if I still can pass an object literal as an argument. 
 
> Having a class instead of interface gives you the benefit of being able to create an object, with faster error detection and clearer intelisense benefits of TS language.

You have exactly the same intellisense language support for interfaces. TS will not allow you to pass an object that doesn't implement the interface. How does creating an object give you a "faster error detection"? 

TypeScript objects are structurally typed - that's why I can pass a literal object as argument to that function. I guess my main issue here is with the constructor. I'm a bit concerned that it may encourage less readable code. Compare these two:

```typescript
import {TemplatesPushRequest} from 'postmark'
client.pushTemplates(new TemplatesPushRequest(0, 1, false))
```

and 

```typescript
client.pushTemplates({SourceServerID: 1, DestinationServerID: 2, PerformChanges: false})
```

Implementing APIs is tricky :)

Reply:
hey @tomazy , bit busy now, so will share details later.

Regarding that argument, you can do:

client.pushTemplates({SourceServerID: 1, DestinationServerID: 2, PerformChanges: false})

with no problems when using class too and you will get proper errors with it.
So I am not sure about that strong opinion about it.

Reply:
check out couple of subtle, but visible differences in intellisense between interface and class.

![screenshot](https://user-images.githubusercontent.com/14041/50225072-cb7a8b00-039f-11e9-9f72-10b437264f25.gif)

When using interfaces, you would get an error on type, but just for first found type error - it acts same way when object on the fly is inside a method. Also auto complete would work only if there is a letter in type in interface. If it's not there you would not be able to autocomplete as easy, which makes constructor descriptions valuable.

Most of it is subtle, but classes are more descriptive, plus you could have other benefits if needed like default values etc - not used in this call though. If you don't need class, thats fine too, TS is versatile, and it would act same way as interface.

It doesn't remove any functionality it just adds to it. 

Thats why our intention was to for all calls that require object creation provide real TS object creation, which is through the class. For other calls, interfaces that only describe types are used.

Reply:
Intellisense is only important when you type this code. You will read it much more often. 

Consider reading your test case in a couple of months and trying to understand what we actually pass to `pushTemplates`:

```js
let pushRequest = new TemplatesPushRequest(0, 1, false)
```
What is `0` - is it the destination or the source server? `false` what?? 

Moreover, JavaScript is not a classical OO language and using classes (and `new`) for value objects is not really idiomatic (it may sometimes matter for performance reasons for objects with methods, but not here).
 
If that's was a part of some private code then it wouldn't matter that much. But here we are suggesting that this is how the library should be used by our clients and to me it's much more problematic. It adds unnecessary ceremony (importing a class, calling `new`, argument order coupling in constructor) whereas an interface would clearly describe a value object (`options`) that is necessary for this function call. 

> Regarding that argument, you can do:
>
> client.pushTemplates({SourceServerID: 1, DestinationServerID: 2, PerformChanges: false})

That's my point! You don't need a class - you need an interface.

Reply:
Well.. class is just an implemented interface if you put it that way.. :)

Alrighty,  I am against using interface for objects needed to be created in TS. We had a long talk about this with @atheken and it made perfect sense after long thinking about it. He definitely won me over, and it's a sensible thing to do for typed languages (main point of TS).

I don't think it makes sense to go back to interfaces for this. If others on team are pro updating classes to interfaces for requests, I am more than willing to make that change.

Reply:
We will continue to define classes for anything that a user needs to create and pass into methods. 

Different editors provide different support, but using classes provides broader support and enables more styles of coding.

Interface definitions are great, and for anything we're returning from our lib, these should be provided as interfaces, not concrete types. 

Some editors will require them to remember the names of the members, having defaults in a constructor simplifies this process and makes it clear exactly which parameters are required vs. optional.

Pure JS consumers also don't have to care about this distinction and can pass an object literal if they choose to. In fact, Typescript users can also pass object literals if they match the interface defined by classes, but defining them as classes allows us to give more editor support for creating them.